volumes/vds/py/vds_functions.py

```diff
@@ -0,0 +1,464 @@
+SLACK_CONN_ID = 'slack_data_pipeline'
+
+def task_fail_slack_alert(dag_name, owners, context):
+    # connection to slack
+    global SLACK_CONN_ID
+    
+    slack_ids = Variable.get('slack_member_id', deserialize_json=True)
+    list_names = []
+    for name in names:
+        list_names.append(slack_ids.get(name, '@Unknown Slack ID')) #find slack ids w/default = Unkown
+
+    slack_webhook_token = BaseHook.get_connection(SLACK_CONN_ID).password
+    
+    log_url = context.get('task_instance').log_url.replace(
+        'localhost', context.get('task_instance').hostname + ":8080"
+    )
+    
+    slack_msg = """
+        :ring_buoy: {dag}.{task} Task Failed.         
+        *Log Url*: {log_url}
+        {slack_name} please check.
+        """.format(
+            task=context.get('task_instance').task_id,
+            dag=context.get('task_instance').dag_id,
+            exec_date=context.get('execution_date'),
+            log_url=log_url,
+            slack_name=' '.join(list_names)
+    )
+    
+    failed_alert = SlackWebhookOperator(
+        task_id='slack_test',
+        http_conn_id='slack',
+        webhook_token=slack_webhook_token,
+        message=slack_msg,
+        username='airflow',
+        proxy='http://'+BaseHook.get_connection('slack').password+'@137.15.73.132:8080',
+        )
+    return failed_alert.execute(context=context)
+
+def fetch_pandas_df(conn, query, table_name):
+    #generic function to pull data in pandas dataframe format 
+    try:
+        with conn.get_conn() as con:
+            with con.cursor() as cur:
+                LOGGER.info(f"Fetching {table_name}")
+                cur.execute(query)
+                data = cur.fetchall()
+                LOGGER.info(f"Number of rows fetched from {table_name} table: {len(data)}")
+                data = pd.DataFrame(data)
+                data.columns=[x.name for x in cur.description]
+                return data
+    except Error as exc:
+        LOGGER.critical(f"Error fetching from {table_name}.")
+        LOGGER.critical(exc)
+        con.close()
+
+def fetch_and_insert_data(select_conn, insert_conn, select_query, insert_query, table_name, batch_size = 100000):
+    #generic function to pull and insert data using different connections and queries.
+    batch=1
+    try:
+        with select_conn.get_conn() as con:
+            with con.cursor() as cur:
+                LOGGER.info(f"Fetching {table_name}")
+                cur.execute(select_query)
+                data = cur.fetchmany(batch_size)
+                while not len(data) == 0:
+                    LOGGER.info(f"Batch {batch} -- {len(data)} rows fetched from {table_name}.")
+                    insert_data(insert_conn, insert_query, table_name, data)
+                    data = cur.fetchmany(batch_size)
+                    batch = batch + 1
+    except Error as exc:
+        LOGGER.critical(f"Error fetching from {table_name}.")
+        LOGGER.critical(exc)
+        con.close()
```

Comment:
`with` should close the connection if there's any error. I don't think you need to close it here